### ISSUE TITLE
Does not work with grouped projects

### ISSUE THREAD

**Issue by teemu-anttila-arek** [mvisonneau/gitlab-ci-pipelines-exporter#210]
This does not fetch the jobs of grouped projects at all.

The /api/v4/projects/ppp/pipelines/nnn gives an expected response if there are no groups on the project hierarchy.

However, if ppp is of format group/subgroup/project, the call fails with 404. Either the slashes on the hierarchy should be URL encoded to %2F or the project number should be used on the call.

**Comment by mvisonneau** [comment]
I do not understand your issue, can you please expand? 🤔

eg: "When I do `x`, I expect `y` but I get `z`"

**Comment by teemu-anttila-arek** [comment]
When I run gitlab-ci-pipelines-exporter against a GitLab account that has projects defined into a group (as we do have all of them), I expect to get the pipeline data read, but I see errors of the API call on the log:

The correct API should be https://xxx.gitlab.xyz/api/v4/projects/group-name%2Fproject-name/pipelines
```
{"level":"info","msg":"discovered new ref","project-name":"group-name/project-name","ref":"master","ref-kind":"branch","time":"2021-01-07T15:40:05Z"}
{"error":"could not read content of pipeline group-name/project-name - master | GET https://xxx.gitlab.xyz/api/v4/projects/group-name/project-name/pipelines/18723: 502 failed to parse unknown error format","level":"warning","msg":"pulling ref metrics","project-name":"group-name/project-name","ref":"master","time":"2021-01-07T16:00:22Z"}
{"error":"error fetching project pipelines for group-name/project-name: error listing project pipelines for project group-name/project-name: GET https://xxx.gitlab.xyz/api/v4/projects/group-name/project-name/pipelines: 502 failed to parse unknown error format","level":"warning","msg":"pulling ref metrics","project-name":"group-name/project-name","ref":"master","time":"2021-01-08T04:46:22Z"}
```

**Comment by mvisonneau** [comment]
I see, interesting, which version of GitLab are you running upon?

**Comment by teemu-anttila-arek** [comment]
GitLab is latest 13.7.2-ee (55a11a0db46) and exporter on v0.4.6 from helm chart gitlab-ci-pipelines-exporter-0.0.9.

**Comment by mvisonneau** [comment]
Hmm this is very strange I have the exact same versions and I did not manage to replicate the error 🤔

Could you provide a sanitized version of your config file?
Also, can you attempt to fetch the URLs with the token you are using to validate it has sufficient permissions?

**Comment by teemu-anttila-arek** [comment]
Very sanitized :)
```
config:
  gitlab:
    token: MASKED
    url: https://xxx.gitlab.xyz
  wildcards:
  - search: xxx
```

The token has access - tested with curl. The API works if the slash between the group name and project name is replaced with %2F.

**Comment by mvisonneau** [comment]
Thanks! I made some further investigation, it seems that the url being used are correct, eg with my load balancer logs:

```
52.x.x.x - - [08/Jan/2021:11:18:58 +0000] "GET /api/v4/projects/group%2Fsubgroup%2Fproject/pipelines/13000 HTTP/1.1" 200 1018 "-" "gitlab-ci-pipelines-exporter-0.4.6" 232 0.049 [gitlab-gitlab-webservice-default-8181] [] 172.x.x.x:8181 1018 0.052 200 34079c50ffed856ad31fecbec03ebb49 -
```

Checking further with cURL, if the API url is not encoded correctly the API returns a `404`, not a `502`

Have you tested the specific `GET https://xxx.gitlab.xyz/api/v4/projects/group-name%2Fproject-name/pipelines/<id>` call with your token? 🤔

**Comment by mvisonneau** [comment]
actually you would get a `401` or `404` as well I believe

**Comment by teemu-anttila-arek** [comment]
Yes, I have tested it and it works fine. Also, replacing group-name/project-name with the project number gives a valid json output.
Your log actually looks like what it should be...

**Comment by anders-ni** [comment]
I am experiencing this same issue. I set up this metrics collection back on tag 0.3.3 and noticed last month that the metrics were no longer pulling. After investigating, I think the metrics stopped pulling after a GitLab upgrade, but I am not sure which version. (I hadn't checked in on the metrics lately)

I upgraded to mvisonneau/gitlab-ci-pipelines-exporter:v0.4.6 to verify that it still did not work with the latest. 

The error I am seeing, when using an empty wild card search to pull all projects is a 404 from the GitLab API
```
{"error":"GET https://xxxx/api/v4/projects/xxgroupxx/xxxproject: 404 {error: 404 Not Found}","level":"warning","msg":"pulling refs from project","project-name":"xxgroupxx/xxxproject","time":"2021-01-13T17:22:16Z"}

{"error":"GET https://xxxx/api/v4/projects/xxgroupxx/xxxproject: 404 {error: 404 Not Found}","level":"warning","msg":"pulling projects refs from pipelines","project-name":"xxgroupxx/xxxproject","time":"2021-01-13T17:22:16Z"}
```
Using the wildcard config to get all projects:
```
wildcards:
    - # Define the owner of the projects we want to look for (optional)
      # Search expression to filter out projects
      # (optional, default: '' -- no filter/all projects)
      search: ''
```

I know the access key being supplied has access to the projects, and I am able to reproduce in a local browser with my own user creds, entering 
```
https://xxxx/api/v4/projects/xxgroupxx%2Fxxxproject or 
https://xxxx/api/v4/projects/xxprojectid 
```
returns the project's response successfully.

However, if I try it with a slash like the code above references, response is the same 404 at the code is experiencing.
`https://xxxx/api/v4/projects/xxgroupxx/xxxproject`.

I have a feeling something changed with the GitLab API in version 13.5, .6 or .7 which made this URL no longer process in the same way on their side. I ran across unexpected changes in other api calls in other places that I was using them, so its plausible something they did changed how these endpoints utilized by your project behave as well.  

Great project, btw. This has enabled a lot of insights into our pipeline metrics we were looking for.

**Comment by belaytzev** [comment]
I have same problem

```
Gitlab v13.7.3
mvisonneau/gitlab-ci-pipelines-exporter v0.4.6
```

```
gitlab:
  url: https://gitlab.example.com
  token: 'Xxx-Xxx'
wildcards:
- owner:
    include_subgroups: true
    kind: group
    name: my-group
```

```
level=debug msg="reading project" project-name=my-group/my-sub-group/my-project
level=warning msg="pulling refs from project" error="GET https://gitlab.example.com/api/v4/projects/my-group/my-sub-group/my-project: 404 {error: 404 Not Found}" project-name=my-group/my-sub-group/my-project
```

**Comment by mvisonneau** [comment]
Thank you @anders-ni  and @belaytzev for sharing further details! I still haven't been able to reproduce using the same versions of GitLab / exporter though 🤔 

I have been checking the SDK and believe the error may be misleading as it may not refer the URL encoded version of the call being made.

I am not convinced this is related to wildcard as the error seems to occur on a project pull. Could you please confirm that if you configure `my-group/my-sub-group/my-project` manually in `projects:` you do not have the error?

**Comment by mvisonneau** [comment]
not sure if any of you may be concerned with: https://stackoverflow.com/questions/63073924/gitlab-api-v4-returns-404-on-project-search-when-using-namespace-projectname-ins 🤔 ?

**Comment by teemu-anttila-arek** [comment]
Using the default nginx that comes with the GitLab helm chart and the default values in front.

**Comment by anders-ni** [comment]
Yes, I get the same 404 error when making api calls for that project if I specify:
```
projects:
    - # Name of the project (actually path with namespace) to fetch
      # (required)
      name: xxxxgroup/xxxxproject
```

I'll check in with someone else on my team that's more familiar with our GitLab install to we have to see if they have any insight into the routing setup. There was a rebuild to our GitLab configuration since I've installed this project initially, there's potential that is where this started failing for me, and not during a version upgrade.

**Comment by mvisonneau** [comment]
Thanks for the confirmation @anders-ni, I appreciate! 🙇 

I believe the original issue that @teemu-anttila-arek raised is related but not exactly the same though. I opened a new one for the 404 you both get @anders-ni  & @belaytzev: https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/issues/214
